libuavcan/include/uavcan/transport/can_io.hpp

```diff
@@ -36,87 +39,83 @@ struct UAVCAN_EXPORT CanRxFrame : public CanFrame
-    struct Entry : public LinkedListNode<Entry>  // Not required to be packed - fits the block in any case
-    {
-        MonotonicTime deadline;
-        CanFrame frame;
-        uint8_t qos;
-        CanIOFlags flags;
+    MonotonicTime deadline;
+    const CanFrame frame;
+    uint8_t qos;
+    CanIOFlags flags;
 
-        Entry(const CanFrame& arg_frame, MonotonicTime arg_deadline, Qos arg_qos, CanIOFlags arg_flags)
-            : deadline(arg_deadline)
+    CanTxQueueEntry(const CanFrame& arg_frame, const MonotonicTime arg_deadline, Qos arg_qos, CanIOFlags arg_flags)
+            : deadline(std::move(arg_deadline))
             , frame(arg_frame)
             , qos(uint8_t(arg_qos))
             , flags(arg_flags)
-        {
-            UAVCAN_ASSERT((qos == Volatile) || (qos == Persistent));
-            IsDynamicallyAllocatable<Entry>::check();
-        }
+    {
+        UAVCAN_ASSERT((qos == Volatile) || (qos == Persistent));
+        IsDynamicallyAllocatable<CanTxQueueEntry>::check();
+    }
 
-        static void destroy(Entry*& obj, IPoolAllocator& allocator);
+    static void destroy(CanTxQueueEntry*& obj, IPoolAllocator& allocator);
 
-        bool isExpired(MonotonicTime timestamp) const { return timestamp > deadline; }
+    bool isExpired(const MonotonicTime timestamp) const { return timestamp > deadline; }
 
-        bool qosHigherThan(const CanFrame& rhs_frame, Qos rhs_qos) const;
-        bool qosLowerThan(const CanFrame& rhs_frame, Qos rhs_qos) const;
-        bool qosHigherThan(const Entry& rhs) const { return qosHigherThan(rhs.frame, Qos(rhs.qos)); }
-        bool qosLowerThan(const Entry& rhs)  const { return qosLowerThan(rhs.frame, Qos(rhs.qos)); }
+    bool operator<(const CanTxQueueEntry & other) const
+    {
+        return this->frame.priorityLowerThan(other.frame);
+    }
 
-#if UAVCAN_TOSTRING
-        std::string toString() const;
-#endif
-    };
+    bool operator>(const CanTxQueueEntry & other) const
+    {
+        return this->frame.priorityHigherThan(other.frame);
+    }
 
-private:
-    class PriorityInsertionComparator
+    bool operator==(const CanTxQueueEntry & other) const
     {
-        const CanFrame& frm_;
-    public:
-        explicit PriorityInsertionComparator(const CanFrame& frm) : frm_(frm) { }
-        bool operator()(const Entry* entry)
-        {
-            UAVCAN_ASSERT(entry);
-            return frm_.priorityHigherThan(entry->frame);
-        }
-    };
+        return this->frame == other.frame;
+    }
 
-    LinkedListRoot<Entry> queue_;
-    LimitedPoolAllocator allocator_;
+#if UAVCAN_TOSTRING
+    std::string toString() const;
+#endif
+};
+
+class UAVCAN_EXPORT CanTxQueue : public AvlTree<CanTxQueueEntry>
+{
+protected:
     ISystemClock& sysclock_;
     uint32_t rejected_frames_cnt_;
 
-    void registerRejectedFrame();
+    bool linkedListContains(Node *head, const CanFrame &frame) const;
+    void safeIncrementRejectedFrames();
+    AvlTree::Node *searchForNonExpiredMax(Node *n);
```

Comment:
Here and elsewhere, please call your asterisk to order. They belong to the type rather than identifier.